bitcoin/src/blockdata/script.rs

```diff
@@ -1311,22 +1342,87 @@ impl ScriptBuf {
     #[inline]
     pub fn into_boxed_script(self) -> Box<Script> {
         // Copied from PathBuf::into_boxed_path
-        let rw = Box::into_raw(self.0.into_boxed_slice()) as *mut Script;
+        let rw = Box::into_raw(self.1.into_boxed_slice()) as *mut Script;
         unsafe { Box::from_raw(rw) }
     }
 }
 
+/// A [`ScriptBuf`] used as the script code when spending segwit utxos.
+///
+/// Please note, the script code created here does not contain the length byte, this is pre-pended
+/// during serialization by `consensus_encode` and in the sighash cache we use `consensus_encode` to
+/// encode the script codes created below - therefore the length byte is not present in the raw
+/// script returned here. To get a hex string use `crate::consensus::serialize_hex(&script_code)`.
+impl ScriptBuf<ScriptCode> {
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output.
+    pub fn p2wpkh_script_code(wpkh: &Script) -> Option<Self> {
```

Comment:
I don't think you can make script code from a script code, the scripts are not recursive. This should be only available for `P2Wpkh` tag. (Or even for `CompressedPublicKey` so people can avoid creating the intermediary script.)

Reply:
Yes you are correct, @sanket1729 told me this but I missed his comment today while working on this. I'm hoping to implement using `WPubkeyHash` (had an initial go at it before but got some grief in `psbt` code). Will do tomorrow.